lib/internal/modules/esm/fetch_module.js

```diff
@@ -0,0 +1,286 @@
+'use strict';
+const {
+  ArrayPrototypePush,
+  Promise,
+  PromisePrototypeThen,
+  PromiseResolve,
+  SafeMap,
+  StringPrototypeEndsWith,
+  StringPrototypeSlice,
+  StringPrototypeStartsWith,
+} = primordials;
+const {
+  Buffer: {
+    concat: BufferConcat
+  }
+} = require('buffer');
+const {
+  ERR_NETWORK_IMPORT_DISALLOWED,
+  ERR_NETWORK_IMPORT_BAD_RESPONSE,
+} = require('internal/errors').codes;
+const { URL } = require('internal/url');
+const net = require('net');
+const path = require('path');
+
+/**
+ * @typedef CacheEntry
+ * @property {Promise<string> | string} resolvedHREF
+ * @property {Record<string, string>} headers
+ * @property {Promise<Buffer> | Buffer} body
+ */
+
+/**
+ * Only for GET requests, other requests would need new Map
+ * HTTP cache semantics keep diff caches
+ *
+ * Maps HREF to pending cache entry
+ * @type {Map<string, Promise<CacheEntry> | CacheEntry>}
+ */
+const cacheForGET = new SafeMap();
+
+// [1] The V8 snapshot doesn't like some C++ APIs to be loaded eagerly. Do it
+// lazily/at runtime and not top level of an internal module.
+
+// [2] Creating a new agent instead of using the gloabl agent improves
+// performance and precludes the agent becoming tainted.
+
+let HTTPSAgent;
+function HTTPSGet(url, opts) {
+  const https = require('https'); // [1]
+  HTTPSAgent ??= new https.Agent({ // [2]
+    keepAlive: true
+  });
+  return https.get(url, {
+    agent: HTTPSAgent,
+    ...opts
+  });
+}
+
+let HTTPAgent;
+function HTTPGet(url, opts) {
+  const http = require('http'); // [1]
+  HTTPAgent ??= new http.Agent({ // [2]
+    keepAlive: true
+  });
+  return http.get(url, {
+    agent: HTTPAgent,
+    ...opts
+  });
+}
+
+function dnsLookup(name, opts) {
+  // eslint-disable-next-line no-func-assign
+  dnsLookup = require('dns/promises').lookup;
+  return dnsLookup(name, opts);
+}
+
+let zlib;
+function createBrotliDecompress() {
+  zlib ??= require('zlib'); // [1]
+  // eslint-disable-next-line no-func-assign
+  createBrotliDecompress = zlib.createBrotliDecompress;
+  return createBrotliDecompress();
+}
+
+function createUnzip() {
+  zlib ??= require('zlib'); // [1]
+  // eslint-disable-next-line no-func-assign
+  createUnzip = zlib.createUnzip;
+  return createUnzip();
+}
+
+/**
+ * @param {URL} parsed
+ * @returns {Promise<CacheEntry> | CacheEntry}
+ */
+function fetchWithRedirects(parsed) {
+  const existing = cacheForGET.get(parsed.href);
+  if (existing) {
+    return existing;
+  }
+  const handler = parsed.protocol === 'http:' ? HTTPGet : HTTPSGet;
+  const result = new Promise((fulfill, reject) => {
+    const req = handler(parsed, {
+      headers: {
+        Accept: '*/*'
+      }
+    })
+    .on('error', reject)
+    .on('response', (res) => {
+      function dispose() {
+        req.destroy();
+        res.destroy();
+      }
+      if (res.statusCode >= 300 && res.statusCode <= 303) {
```

Comment:
We should probably think of adding support for HTTP status code 307 and 308 at some point.